tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc

```diff
@@ -602,22 +602,91 @@ CudnnConvParamsKey makeConvTuningCacheKey(MemRefType<T, N>& input,
                                           MemRefType<int32_t, 1>& paddings,
                                           MemRefType<T, N>& output,
                                           MemRefType<int32_t, 1>& metadata) {
+  // Note metadata order:
+  //   - input layout: each field for one dimension. The order is:
+  //     * batch, channel, spatial dimensions
+  //   - kernel layout: each field for one dimension. The order is:
+  //     * in_channel, out_channel, spatial dimensions
+  //   - output layout: each field for one dimension. The order is:
+  //     * batch, channel, spatial dimensions
+  //   - strides: each filed for one spatial dimension.
+  //   - dilations: each filed for one spatial dimension.
   CudnnConvParamsKey key;
-  key.input_shape.reserve(N);
-  key.filter_shape.reserve(N);
-  key.output_shape.reserve(N);
-  for (int i = 0; i < N; ++i) {
-    key.input_shape.push_back(input.sizes[i]);
-    key.filter_shape.push_back(filter.sizes[i]);
-    key.output_shape.push_back(output.sizes[i]);
+  static_assert(N > 2);
+  const int effectiveN = (N < 4) ? 4 : N;
+  std::vector<int64_t> input_sizes(effectiveN);
+  std::vector<int64_t> filter_sizes(effectiveN);
+  std::vector<int64_t> output_sizes(effectiveN);
+  int64_t padding_orig_sizes = paddings.sizes[0];
+  std::vector<int32_t> padding_data(effectiveN);
+  int64_t metadata_orig_sizes = metadata.sizes[0];
+  std::vector<int32_t> metadata_data;
+
+  bool is_1d_conv = (N == 3);
+  // input, kernel, and output layout & sizes.
+  for (int64_t layout_idx = 0; layout_idx < 3; layout_idx++) {
+    std::vector<int64_t>* sizes;
+    int64_t* orig_sizes;
+    if (layout_idx == 0) {
+      sizes = &input_sizes;
+      orig_sizes = &(input.sizes[0]);
+    } else if (layout_idx == 1) {
+      sizes = &filter_sizes;
+      orig_sizes = &(filter.sizes[0]);
+    } else {
+      sizes = &output_sizes;
+      orig_sizes = &(output.sizes[0]);
+    }
+    int64_t metadata_idx_base = layout_idx * N;
+    int64_t spatial_dim = metadata.data[metadata_idx_base + 2];
+    for (int64_t i = 0; i < N; i++) {
+      int64_t idx = metadata_idx_base + i;
+      int64_t orig_dim = metadata.data[idx];
+      int64_t dim = orig_dim;
+      if (is_1d_conv && (orig_dim > spatial_dim)) {
+        // We will insert one spatial dim just after the existing spatial dim.
```

Comment:
nit: We usually expand the one dim right before the original spatial dim (like we do on disc-cpu or the logic in TensorFlow). Do you see any performance difference these two strategies?

Reply:
The new commit inserts the extra dim just before the existing spatial dim. No performance difference for the case I tested.